docs/development.md

```diff
@@ -56,6 +56,139 @@ git push -f origin myfeature
+
+4. Use https://godoc.org/k8s.io/client-go/testing#Fake.Actions 
+  for validating number of fake actions performed and the values with which the fake calls were made.
+
+##### Example of using fake for testing functions making API calls.
+
+Initialising a fakeclientset and fakeclient properly is the first thing. 
+
+For example : 
+take the example of unit test for CreateRoute function in pkg/occlient/occlient.go 
+```
+/ CreateRoute creates a route object for the given service and with the given
+// labels
+func (c *Client) CreateRoute(name string, serviceName string, labels map[string]string) (*routev1.Route, error) {
+	route := &routev1.Route{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   name,
+			Labels: labels,
+		},
+		Spec: routev1.RouteSpec{
+			To: routev1.RouteTargetReference{
+				Kind: "Service",
+				Name: serviceName,
+			},
+		},
+	}
+	r, err := c.routeClient.Routes(c.namespace).Create(route)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating route")
+	}
+	return r, nil
+}
+```
+
+Looking at the function body and identify how many API calls it is making while execution. 
+In this case CreateRoute is making only a single API call which is 
+```
+r, err := c.routeClient.Routes(c.namespace).Create(route)
+```
+
+for example,
+add the code for initialising fakeclientset & fakeclient for routeClient on the FakeNew function 
+```
+    fkclientset.RouteClientset = fakeRouteClientset.NewSimpleClientset()
+    client.routeClient = fkclientset.RouteClientset.Route()
+
+```
+Initialise a fakeclientset by calling fakeRouteClientset.NewSimpleClientset() 
+It returns a simple set of object tracker which can process, creates,updates and deletions 
+but without any validations. So it is recommended to implement validations separately if needed.
+
+Writing the test function for CreateRoute,
+
+call the initialisation function
+```
+fkclient, fkclientset := FakeNew()
+```
+
+Once fakeclientset and fakeclient is being sucessfully initialised 
+it's possible to make the method calls on the struct which got initialised using FakeNew().
+Here in this case it needs to validate the value with which the function is returining
+and the number of actions performed on routeclientset.
+
+
+Then after making that function call inside the test function,
+the number of actions performed can be validated using being validated Actions() method on clientset.
+
+for example in this case, `len(fkclientset.RouteClientset.Actions())` 
+returns number of actions performed on routeclientset
+
+and values are being validated later.
+Ref. (here)[https://github.com/redhat-developer/odo/pull/456/files#diff-54c1e3725d2cfb565cbd1cfdb02bd792R63]
+
+For the API calls which are returning objects that are later being processed inside the function body,
+adding reactors for the relevent actions is also needed.
+
+for example take a look at `GetDeploymentConfigsFromSelector` 
```

Comment:
I think it would be nice to convert all(as much as possible) all github links to markdown links and not expose the github urls(Github urls seem very ugly to me from the perspective of a reader :) but yes this for a special reader -- A DEVELOPER :P but still I feel it would be nice to avoid it)

Reply:
If possible can we pick one single example that covers all aspects of testing and only consider that throughout the doc?
Not a very crucial thing to be done, but I feel it avoid context switches in reader's mind